src/main.ts

```diff
@@ -0,0 +1,115 @@
+/* eslint-disable no-console, @typescript-eslint/no-explicit-any */
+export const LogLevels = {
+  DEBUG: 10,
+  INFO: 20,
+  WARNING: 30,
+  ERROR: 40,
+};
+
+type Tags = Record<string, any>;
+
+export class ProperLogger {
+  name: string;
+
+  commonTags: Tags;
+
+  logLevel: number;
+
+  console: Pick<Console, 'debug' | 'log' | 'warn' | 'error'>;
+
+  static getLogger(name: string): ProperLogger {
+    return new ProperLogger(name);
+  }
+
+  constructor(name: string) {
+    this.name = name;
+    // Here we grab a reference to the `global.console` so that in tests we can
+    // patch this out.
+    this.console = global.console;
+    this.commonTags = {};
+    this.logLevel = LogLevels.INFO;
+    if (process.env.PROPERLY_LOG_LEVEL) {
+      const envValue = parseInt(process.env.PROPERLY_LOG_LEVEL, 10);
+      if (envValue >= 0 && envValue < Number.MAX_SAFE_INTEGER) {
+        this.logLevel = envValue;
+      }
+    }
+  }
+
+  addCommonTags(commonTags: Tags): void {
+    this.commonTags = {
+      ...this.commonTags,
+      ...commonTags,
+    };
+  }
+
+  clearCommonTags(): void {
+    this.commonTags = {};
+  }
+
+  private prepareLog(
+    message: string | null,
+    thingsToLog: Record<string, any>
+  ): string {
+    const firstPart: Record<string, string> = {};
+    if (message) {
+      firstPart.message = message;
+    }
+
+    return JSON.stringify({ ...firstPart, ...this.commonTags, ...thingsToLog });
+  }
+
+  debug(message: string, extraTags?: Tags): void {
+    if (this.logLevel > LogLevels.DEBUG) {
+      return;
+    }
+    this.console.debug(this.prepareLog(message, { ...extraTags }));
```

Comment:
Do we need to copy `extraTags` (vs. passing it directly)?
(all logging methods do this and I can't see why offhand)

Reply:
I take that back. I tried making the `extraTags: Tags | undefined` (this is slightly different from `extraTags?: Tags` because it's a required argument with an optional value rather than an optional argument), but I didn't like it (no particular reason). The `|| {}` "felt nicer".